File: spi/src/main/java/com/github/ykiselev/spi/camera/Plane.java
package com.github.ykiselev.spi.camera;

import com.github.ykiselev.opengl.matrices.Vector3f;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public final class Plane {

    public enum Classification {
        INSIDE, OUTSIDE, ON_PLANE
    }

    private final Logger logger = LoggerFactory.getLogger(getClass());

    private final Vector3f normal = new Vector3f();

    private float d;

    public void set(float a, float b, float c, float d) {
        normal.set(a, b, c);
        double ool = normal.normalize();
        this.d = (float) (d * ool);
    }

    public Classification classify(Vector3f p) {
        return classify(p, 0f);
    }

    public Classification classify(Vector3f p, float radius) {
        double r = normal.dotProduct(p) + d + radius;
        if (r < 0) {
            logger.debug("{} with radius {} is outside (by {}) of {}", p, radius, r, this);
            return Classification.OUTSIDE;
        } else if (r == 0) {
            return Classification.ON_PLANE;
        }
        return Classification.INSIDE;
    }

    @Override
    public String toString() {
        return "Plane{" +
                "normal=" + normal +
                ", d=" + d +
                '}';
    }
}

File: spi/src/main/java/com/github/ykiselev/spi/camera/Plane.kt
package com.github.ykiselev.spi.camera

import com.github.ykiselev.opengl.matrices.Vector3f

class Plane {
    enum class Classification {
        INSIDE, OUTSIDE, ON_PLANE
    }

    val normal = Vector3f()

    var d: Float = 0f
        private set

    fun set(a: Float, b: Float, c: Float, d: Float) {
        val n = Vector3f(a, b, c)
        val len = n.length()
        check(len != 0.0) {
            "Invalid plane: ($n, $d)"
        }
        val ool = 1.0 / len
        normal.set((n.x * ool).toFloat(), (n.y * ool).toFloat(), (n.z * ool).toFloat())
        this.d = (d * ool).toFloat()
    }

    fun classify(p: Vector3f, radius: Float = 0f): Classification {
        val r = normal.dot(p) + d + radius
        if (r < 0) {
            return Classification.OUTSIDE
        } else if (r == 0.0) {
            return Classification.ON_PLANE
        }
        return Classification.INSIDE
    }
}
